packages/ember-application/tests/system/visit_test.js

```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
 
-  QUnit.test('visit() returns a promise that resolves when the view has rendered', function(assert) {
-    QUnit.expect(3);
+  QUnit.test('visit() rejects on application boot failure', function(assert) {
+    QUnit.expect(2);
+    QUnit.stop();
+
+    run(function() {
+      createApplication();
+
+      App.initializer({
+        name: 'error',
+        initialize() {
+          throw new Error('boot failure');
+        }
+      });
+
+      App.visit('/').then(
+        () => {
+          QUnit.start();
+          assert.ok(false, 'It should not resolve the promise');
+        },
+        (error) => {
+          QUnit.start();
+          assert.ok(error instanceof Error, 'It should reject the promise with the boot error');
+          assert.equal(error.message, 'boot failure');
+        }
+      );
+    });
+  });
+
+  QUnit.test('visit() rejects on instance boot failure', function(assert) {
+    QUnit.expect(2);
     QUnit.stop();
 
-    var app;
+    run(function() {
+      createApplication();
+
+      App.instanceInitializer({
+        name: 'error',
+        initialize() {
+          throw new Error('boot failure');
+        }
+      });
+
+      App.visit('/').then(
+        () => {
+          QUnit.start();
+          assert.ok(false, 'It should not resolve the promise');
+        },
+        (error) => {
+          QUnit.start();
+          assert.ok(error instanceof Error, 'It should reject the promise with the boot error');
+          assert.equal(error.message, 'boot failure');
+        }
+      );
+    });
+  });
+
+  QUnit.test('visit() follows redirects', function(assert) {
+    QUnit.expect(2);
+    QUnit.stop();
 
     run(function() {
-      app = createApplication();
-      app.instanceInitializer({
-        name: 'register-application-template',
-        initialize(app) {
-          app.register('template:application', compile('<h1>Hello world</h1>'));
+      createApplication();
+
+      App.Router.map(function() {
+        this.route('a');
+        this.route('b', { path: '/b/:b' });
+        this.route('c', { path: '/c/:c' });
+      });
+
+      App.register('route:a', Route.extend({
+        afterModel() {
+          this.replaceWith('b', 'zomg');
+        }
+      }));
+
+      App.register('route:b', Route.extend({
+        afterModel(params) {
+          this.transitionTo('c', params.b);
+        }
+      }));
+
+      App.visit('/a').then(
+        (instance) => {
+          QUnit.start();
+          assert.ok(instance instanceof ApplicationInstance, 'promise is resolved with an ApplicationInstance');
+          assert.equal(instance.getURL(), '/c/zomg', 'It should follow all redirects');
+        },
+        (error) => {
+          QUnit.start();
+          assert.ok(false, 'The visit() promise was rejected: ' + error);
+        }
+      );
+    });
+  });
+
+  QUnit.test('visit() rejects if an error occured during a transition', function(assert) {
+    QUnit.expect(2);
+    QUnit.stop();
+
+    run(function() {
```

Comment:
Are we actually intentionally creating and visiting in the same run loop? This does not seem correct, or a natural thing to happen.